### PR TITLE
msrun & protocol model additions

### PR DESCRIPTION
**Pull request by hepcat72**
# TraceBase Pull Request Form

## Summary Change Description

Added MSRun and Protocol tables to the model.

## Resolved Issue Numbers

- #18

## Code Review Notes

Note, I gave the foreign key names the names of the tables they link to instead of the <table_primarykeyfield> convention with our other projects because of the way that the python code accesses the references records, e.g. `msr.protocol.id` refers to the `id` field in the linked protocol record.

## Checklist

- [x] All (linked above) issue requirements satisfied
- [x] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
- [x] [Tests implemented](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
- [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)

### CODE REVIEW

**Comment by hepcat72** [comment]
Looks like I need to edit the behaviors of black and possibly other linters.  Everything passed on the command line...

**Comment by lparsons** [comment]
@hepcat72 PR #25 fixes the linting issues. Feel free to rebase this branch to resolve those issues.

File: .flake8 [hunk 1]
```diff
@@ -2,4 +2,4 @@
 max-line-length = 120
 ignore =
     E231  # There should be whitespace after the characters ,, ;, and :. black; fixed in black 20 (https://github.com/psf/black/issues/1289)
-exclude = .venv
+exclude = .venv,__pycache__,migrations
```
Comment (lparsons) [inline .flake8:5]: Might consider having `migrations`, `.venv` and `__pycache__` igored at the superlinter level to simplify, but that depends on how people are running things locally. I've been running linters in my editor (so these ignores don't matter) and running super linter from Docker as a final check.

To do that, add a line like this to `.github/workflows/superlinter.yml`:
```yaml
          FILTER_REGEX_EXCLUDE: (\.pylintrc|\.venv|__pycache__|migrations)
```

Comment (hepcat72) [reply .flake8:5]: - [ ] Add to `migrations`, `.venv`, and `__pycache__` to `.github/workflows/superlinter.yml` and remove them from the individual linter configs.

**Comment by hepcat72** [comment]
> @hepcat72 PR #25 fixes the linting issues. Feel free to rebase this branch to resolve those issues.

OK.  I approved #25.  I will merge and rebase.

File: .github/workflows/superlinter.yml [hunk 2]
```diff
@@ -24,3 +24,4 @@ jobs:
           DEFAULT_BRANCH: main
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
           LINTER_RULES_PATH: /
+          FILTER_REGEX_EXCLUDE: \.venv,__pycache__,migrations,\.pylintrc
```
Comment (hepcat72) [inline .github/workflows/superlinter.yml:27]: Huh... why didn't this work to avoid the super linter errors about `.pylintrc` and `migrations`...

Comment (hepcat72) [reply .github/workflows/superlinter.yml:27]: Oh, whoops.  I see. `(\.pylintrc|\.venv|__pycache__|migrations)`

**Comment by lparsons** [comment]
> > @hepcat72 PR #25 fixes the linting issues. Feel free to rebase this branch to resolve those issues.
> 
> OK. I approved #25. I will merge and rebase.

Merge or rebase, not both. My preference would be rebase to keep the history clean, but either way is fine.

**Comment by hepcat72** [comment]
> > > @hepcat72 PR #25 fixes the linting issues. Feel free to rebase this branch to resolve those issues.
> > 
> > 
> > OK. I approved #25. I will merge and rebase.
> 
> Merge or rebase, not both. My preference would be rebase to keep the history clean, but either way is fine.

I meant merge PR #25 with main, then rebase off main.  🙄

**Comment by lparsons** [comment]
> > > > @hepcat72 PR #25 fixes the linting issues. Feel free to rebase this branch to resolve those issues.
> > > 
> > > 
> > > OK. I approved #25. I will merge and rebase.
> > 
> > 
> > Merge or rebase, not both. My preference would be rebase to keep the history clean, but either way is fine.
> 
> I meant merge PR #25 with main, then rebase off main. 🙄

Yup, seems obvious now, sorry about that 😕

**Comment by hepcat72** [comment]
> > > > > @hepcat72 PR #25 fixes the linting issues. Feel free to rebase this branch to resolve those issues.
> > > > 
> > > > 
> > > > OK. I approved #25. I will merge and rebase.
> > > 
> > > 
> > > Merge or rebase, not both. My preference would be rebase to keep the history clean, but either way is fine.
> > 
> > 
> > I meant merge PR #25 with main, then rebase off main. 🙄
> 
> Yup, seems obvious now, sorry about that 😕

Lol, don't worry about it.  It was worth the chuckle.

**Review by jcmatese — APPROVED** [review]
Looks fine to me.  Could you please comment on the renaming conventions from sample_id to sample (and protocol_id to protocol)?  I was wondering, because the database attribute itself is a number. If someone were call say `msrun.protocol`, are they going to get an instance of the protocol class or will they get a number (protocol.id)?

**Comment by jcmatese** [comment]
>  Django tutorial memory is not ingrained.

I think I got it.

https://stackoverflow.com/questions/24892148/how-does-django-foreign-key-access-work
https://docs.djangoproject.com/en/3.1/ref/models/fields/#django.db.models.ForeignKey


"Database Representation

Behind the scenes, Django appends "_id" to the field name to create its database column name. In the above example, the database table for the Car model will have a manufacturer_id column. (You can change this explicitly by specifying db_column) However, your code should never have to deal with the database column name, unless you write custom SQL. You’ll always deal with the field names of your model object."

**Comment by hepcat72** [comment]
> > Django tutorial memory is not ingrained.
> 
> I think I got it.
> 
> https://stackoverflow.com/questions/24892148/how-does-django-foreign-key-access-work
> https://docs.djangoproject.com/en/3.1/ref/models/fields/#django.db.models.ForeignKey
> 
> "Database Representation
> 
> Behind the scenes, Django appends "_id" to the field name to create its database column name. In the above example, the database table for the Car model will have a manufacturer_id column. (You can change this explicitly by specifying db_column) However, your code should never have to deal with the database column name, unless you write custom SQL. You’ll always deal with the field names of your model object."

What is it you worked out John? Are you running SQL queries in postgres? If you are, I would recommend instead using a python shell and running queries using python ORM commands.  That's what I did when I noted weirdities with the schema.  I posted about it on Slack (but maybe that was a conversation with Lance). If that's what you're doing, I can find my previous post/comment and paste it here so you can see how to do it...

**Comment by hepcat72** [comment]
BTW, I just resolved a conflict on this branch with main after my rebase yesterday. I hadn't noticed the conflict until now. Preparing to add a couple notes about migrations to the PR form and the CONTRIBUTING doc, and I figure I'll branch off this since it has unmerged CONTRIBUTING changes.

**Comment by hepcat72** [comment]
It's i9n the #development channel on Slack...  Here's a summation example:

```
$ python manage.py shell
>>> from DataRepo.models import Compound
>>> cpd = Compound.objects.get(id=1)
>>> cpd.name
'alanine'
>>> cpd.formula
'C3H7NO2'
>>> cpd.hmdb_id
'HMDB0000161'
```
I had done this because of the weird way Django and postgres fight over capiutalization.

**Comment by jcmatese** [comment]
> What is it you worked out John? 

Getting the tips/cheatsheet for foreign key specification in Django models.  
i.e.
Don't use *_id in your foreign key model specification, because Django does that automatically.
If you _need_ the foreign key value, you could probably retrieve it via with the Django-autogenerated columns name (like msrun.protocol_id)
Otherwise, the attribute retrieves the instantiated class of the foreign key, I think. msrun.protocol returns the instance, not the number/sequence.


> Are you running SQL queries in postgres? 
no, I was just talking about naming conventions and usage for singular foreign keys.

Now, I should probably do that for ManyToMany relationships...

**Review by fkang-pu — APPROVED** [review]

**Review by lparsons — APPROVED** [review]
LGTM! :shipit:

**Pull request merged**